File: java/src/test/java/com/aaronicsubstances/shrewd/evolver/LogRecordFormatParserTest.java
package com.aaronicsubstances.shrewd.evolver;

import java.util.Arrays;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import static java.util.Arrays.asList;
import static com.aaronicsubstances.shrewd.evolver.TestUtils.toMap;

import com.aaronicsubstances.shrewd.evolver.LogRecordFormatParser.FormatTokenType;
import com.aaronicsubstances.shrewd.evolver.LogRecordFormatParser.PartDescriptor;
import static org.testng.Assert.*;

import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

public class LogRecordFormatParserTest {

    public static class TokenPart {
        final PartDescriptor part;
        final boolean isReplacementField;
        final int startPos;
        final int endPos;

        public TokenPart(PartDescriptor part, boolean isReplacementField, int startPos, int endPos) {
            this.part = part;
            this.isReplacementField = isReplacementField;
            this.startPos = startPos;
            this.endPos = endPos;
        }

        @Override
        public boolean equals(Object obj) {
            if (!(obj instanceof TokenPart)) return false;
            TokenPart other = (TokenPart) obj;
            return other.part.equals(part) &&
                other.isReplacementField == isReplacementField &&
                other.startPos == startPos &&
                other.endPos == endPos;
        }

        @Override
        public int hashCode() {
            return Objects.hash(part, isReplacementField, startPos, endPos);
        }

        @Override
        public String toString() {
            return "TokenPart{part=" + part + ", isReplacementField=" + isReplacementField + 
                ", startPos=" + startPos + ", endPos=" + endPos + "}";
        }
    }

    @Test(dataProvider = "createTestCalculateLineAndColumnNumbersData")
    public void testCalculateLineAndColumnNumbers(String s, int pos,
            int expLineNumber, int expColumnNumber) {
        int[] expected = new int[]{ expLineNumber, expColumnNumber };
        int[] actual = LogRecordFormatParser.calculateLineAndColumnNumbers(s, pos);
        assertEquals(actual, expected);
    }

    @DataProvider
    public Object[][] createTestCalculateLineAndColumnNumbersData() {
        String inputWin32 = "This this is \r\nthe GOD we \r\nadore\r\n.";
        String inputUnix = "This this is \nthe GOD we \nadore\n.";
        String inputMac = "This this is \rthe GOD we \radore\r.";

        return new Object[][]{
            { "", 0, 1, 1 },
            { "\n", 1, 2, 1 },
            { "abc", 3, 1, 4 },
            { "ab\nc", 4, 2, 2 },
            { "ab\nc\r\n", 6, 3, 1 },

            new Object[]{ inputWin32, 1, 1, 2 },
            new Object[]{ inputWin32, 4, 1, 5 },
            new Object[]{ inputWin32, 5, 1, 6 },
            new Object[]{ inputWin32, 9, 1, 10 },
            new Object[]{ inputWin32, 10, 1, 11 },
            new Object[]{ inputWin32, 12, 1, 13 },
            new Object[]{ inputWin32, 13, 1, 14 },
            new Object[]{ inputWin32, 14, 1, 15 }, // test that encountering \r does not lead to newline
            new Object[]{ inputWin32, 15, 2, 1 },  // until \n here on this line.
            new Object[]{ inputWin32, 18, 2, 4 },

            new Object[]{ inputUnix, 1, 1, 2 },
            new Object[]{ inputUnix, 4, 1, 5 },
            new Object[]{ inputUnix, 5, 1, 6 },
            new Object[]{ inputUnix, 9, 1, 10 },
            new Object[]{ inputUnix, 10, 1, 11 },
            new Object[]{ inputUnix, 12, 1, 13 },
            new Object[]{ inputUnix, 13, 1, 14 },
            new Object[]{ inputUnix, 14, 2, 1 },
            new Object[]{ inputUnix, 17, 2, 4 },

            new Object[]{ inputMac, 1, 1, 2 },
            new Object[]{ inputMac, 4, 1, 5 },
            new Object[]{ inputMac, 5, 1, 6 },
            new Object[]{ inputMac, 9, 1, 10 },
            new Object[]{ inputMac, 10, 1, 11 },
            new Object[]{ inputMac, 12, 1, 13 },
            new Object[]{ inputMac, 13, 1, 14 },
            new Object[]{ inputMac, 14, 2, 1 },
            new Object[]{ inputMac, 17, 2, 4 },
        };
    }

    @Test(dataProvider = "createTestParseOnePartData")
    public void testParseOnePart(String source, List<TokenPart> expected) {
        LogRecordFormatParser tokenizer = new LogRecordFormatParser(source);
        PartDescriptor part;
        List<TokenPart> actual = new ArrayList<>();
        while ((part = tokenizer.parseOnePart()) != null) {
            assertTrue(asList(FormatTokenType.LITERAL_STRING_SECTION, 
                FormatTokenType.END_REPLACEMENT).contains(tokenizer.tokenType));
            TokenPart partDescriptor = new TokenPart(part,
                tokenizer.tokenType == FormatTokenType.END_REPLACEMENT,
                tokenizer.partStart, tokenizer.endPos);
            actual.add(partDescriptor);
        }
        assertEquals(actual, expected);
    }

    @DataProvider
    public Object[][] createTestParseOnePartData() {
        return new Object[][]{
            { "", asList() },
            { " ", asList(
                new TokenPart(new PartDescriptor(" "), false, 0, 1)
                ) },
            { "xyz", asList(
                new TokenPart(new PartDescriptor("xyz"), false, 0, 3)
                ) },
            { "x{{}}{{yz", asList(
                new TokenPart(new PartDescriptor("x{}{yz"), false, 0, 9)
                ) },
            { "x{}{{yz", asList(
                new TokenPart(new PartDescriptor("x"), false, 0, 1),
                new TokenPart(new PartDescriptor(asList()), true, 1, 3),
                new TokenPart(new PartDescriptor("{yz"), false, 3, 7)
                ) },
            { "x{0}{{yz", asList(
                new TokenPart(new PartDescriptor("x"), false, 0, 1),
                new TokenPart(new PartDescriptor(0), true, 1, 4),
                new TokenPart(new PartDescriptor("{yz"), false, 4, 8)
                ) },
            { "x{$0}{{yz", asList(
                new TokenPart(new PartDescriptor("x"), false, 0, 1),
                new TokenPart(new PartDescriptor(0), true, 1, 5),
                new TokenPart(new PartDescriptor("{yz"), false, 5, 9)
                ) },
            { "x{a}{{yz", asList(
                new TokenPart(new PartDescriptor("x"), false, 0, 1),
                new TokenPart(new PartDescriptor(asList("a")), true, 1, 4),
                new TokenPart(new PartDescriptor("{yz"), false, 4, 8)
                ) },
            { "x{$a}{{yz", asList(
                new TokenPart(new PartDescriptor("x"), false, 0, 1),
                new TokenPart(new PartDescriptor(asList("a"), false), true, 1, 5),
                new TokenPart(new PartDescriptor("{yz"), false, 5, 9)
                ) },
            { "x{.0}{{yz", asList(
                new TokenPart(new PartDescriptor("x"), false, 0, 1),
                new TokenPart(new PartDescriptor(asList("0")), true, 1, 5),
                new TokenPart(new PartDescriptor("{yz"), false, 5, 9)
                ) },
            { "x{$.0}{{yz", asList(
                new TokenPart(new PartDescriptor("x"), false, 0, 1),
                new TokenPart(new PartDescriptor(asList("0"), false), true, 1, 6),
                new TokenPart(new PartDescriptor("{yz"), false, 6, 10)
                ) },

            // test previous 4 again, but with whitespace tolerance.
            { "x{ }{{yz", asList(
                new TokenPart(new PartDescriptor("x"), false, 0, 1),
                new TokenPart(new PartDescriptor(asList()), true, 1, 4),
                new TokenPart(new PartDescriptor("{yz"), false, 4, 8)
                ) },
            { "x{ 0 }{{yz", asList(
                new TokenPart(new PartDescriptor("x"), false, 0, 1),
                new TokenPart(new PartDescriptor(0), true, 1, 6),
                new TokenPart(new PartDescriptor("{yz"), false, 6, 10)
                ) },
            { "x{a  }{{yz", asList(
                new TokenPart(new PartDescriptor("x"), false, 0, 1),
                new TokenPart(new PartDescriptor(asList("a")), true, 1, 6),
                new TokenPart(new PartDescriptor("{yz"), false, 6, 10)
                ) },
            { "x{ . 0 }{{yz", asList(
                new TokenPart(new PartDescriptor("x"), false, 0, 1),
                new TokenPart(new PartDescriptor(asList("0")), true, 1, 8),
                new TokenPart(new PartDescriptor("{yz"), false, 8, 12)
                ) },

            // handle longer replacement fields.
            { "{bag .prices[0 ]}", asList(
                new TokenPart(new PartDescriptor(asList("bag", "prices", 0)), true, 0, 17)
                ) },                
            { "{ -12 }{ [10][-2] }{ y.z }", asList(
                new TokenPart(new PartDescriptor(-12), true, 0, 7),
                new TokenPart(new PartDescriptor(asList(10, -2)), true, 7, 19),
                new TokenPart(new PartDescriptor(asList("y", "z")), true, 19, 26)
                ) }
        };
    }    

    @Test(dataProvider = "createTestParseData")
    public void testParse(int index, String source, List<PartDescriptor> expected) {
        List<PartDescriptor> actual = null;
        try {
            LogRecordFormatParser instance = new LogRecordFormatParser(source);
            actual = instance.parse();
        }
        catch (Throwable ex) {
            System.err.println((index + 1) + ". " + ex);
        }
        assertEquals(actual, expected);
    }

    @DataProvider
    public Object[][] createTestParseData() {
        int index = 0;
        return new Object[][]{
            { index++, "{", null },
            { index++, "}", null },
            { index++, "{0", null },
            { index++, "{a}", asList(new PartDescriptor(asList("a"))) },
            { index++, "a{.0.[2]}b", null },
            { index++, "a{.}b", null },
            { index++, "a{.a$}b", null },
            { index++, "a{.{b", null },
            { index++, "a{.x{b", null },
            { index++, "a{[0}b", null },
            { index++, "a{0]}b", null },
            { index++, "a{.0]}b", null },
            { index++, "a{[]}b", null },
            { index++, "a{[x]}b", null },
            { index++, "a{[200]}b", asList(new PartDescriptor("a"),
                new PartDescriptor(asList(200)), new PartDescriptor("b")) },
            { index++, "a{ $ [ 200]}b", null },
            { index++, "a{$ [ 200]}b", asList(new PartDescriptor("a"),
                new PartDescriptor(asList(200), false), new PartDescriptor("b")) },
            { index++, " a { x.data [ 200 ] [ 300 ] . q . z } b", asList(new PartDescriptor(" a "), 
                new PartDescriptor(asList("x", "data", 200, 300, "q", "z")), 
                new PartDescriptor(" b")) },
            { index++, "", asList() },
            { index++, "{}", asList(new PartDescriptor(asList())) },
            { index++, "{$}", asList(new PartDescriptor(asList(), false)) },
            { index++, ".[]", asList(new PartDescriptor(".[]")) },
            { index++, ".{{{{x}}}}", asList(new PartDescriptor(".{{x}}")) },
            { index++, "0", asList(new PartDescriptor("0")) },
            { index++, "{{0}}", asList(new PartDescriptor("{0}")) },
            { index++, "{0}", asList(new PartDescriptor(0)) },
            { index++, "{$0}", asList(new PartDescriptor(0)) },
            { index++, "{[0]}", asList(new PartDescriptor(asList(0))) },
            { index++, "{$[0]}", asList(new PartDescriptor(asList(0), false)) },

            // test with newlines
            { index++, "{[\n0]\n}", asList(new PartDescriptor(asList(0))) },
            { index++, "\nThere is plenty{\n}\n of peace\n", 
                asList(new PartDescriptor("\nThere is plenty"), new PartDescriptor(asList()),
                    new PartDescriptor("\n of peace\n")) },
            { index++, "{[0]}\nfirst\nsecond{a}}r{y}", null }
        };
    }
}

File: csharp/ShrewdEvolver.UnitTests/LogRecordFormatParserTest.cs
﻿using System;
using System.Collections.Generic;
using System.Text;
using Xunit;
using static AaronicSubstances.ShrewdEvolver.UnitTests.TestUtils;

namespace AaronicSubstances.ShrewdEvolver.UnitTests
{
    public class LogRecordFormatParserTest
    {
        public class TokenPart
        {
            internal object part;
            internal readonly bool isReplacementField;
            internal readonly int startPos;
            internal readonly int endPos;

            public TokenPart(object part, bool isReplacementField, int startPos, int endPos)
            {
                this.part = part;
                this.isReplacementField = isReplacementField;
                this.startPos = startPos;
                this.endPos = endPos;
            }
            
            // override object.Equals
            public override bool Equals(object obj)
            {
                if (obj == null || GetType() != obj.GetType())
                {
                    return false;
                }

                var other = (TokenPart)obj;
                return other.part.Equals(part) &&
                    other.isReplacementField == isReplacementField &&
                    other.startPos == startPos &&
                    other.endPos == endPos;
            }

            // override object.GetHashCode
            public override int GetHashCode()
            {
                return 1; 
            }

            public override string ToString()
            {
                return "TokenPart{part=" + part + ", isReplacementField=" + isReplacementField +
                    ", startPos=" + startPos + ", endPos=" + endPos + "}";
            }
        }

        [Theory]
        [MemberData(nameof(CreateTestCalculateLineAndColumnNumbersData))]
        public void TestCalculateLineAndColumnNumbers(string s, int pos,
            int expLineNumber, int expColumnNumber)
        {
            int[] expected = new int[] { expLineNumber, expColumnNumber };
            int[] actual = LogRecordFormatParser.CalculateLineAndColumnNumbers(s, pos);
            Assert.Equal(expected, actual);
        }

        public static List<object[]> CreateTestCalculateLineAndColumnNumbersData()
        {
            string inputWin32 = "This this is \r\nthe GOD we \r\nadore\r\n.";
            string inputUnix = "This this is \nthe GOD we \nadore\n.";
            string inputMac = "This this is \rthe GOD we \radore\r.";

            return new List<object []>
            {
                new object[]{ "", 0, 1, 1 },
                new object[]{ "\n", 1, 2, 1 },
                new object[]{ "abc", 3, 1, 4 },
                new object[]{ "ab\nc", 4, 2, 2 },
                new object[]{ "ab\nc\r\n", 6, 3, 1 },

                new object[]{ inputWin32, 1, 1, 2 },
                new object[]{ inputWin32, 4, 1, 5 },
                new object[]{ inputWin32, 5, 1, 6 },
                new object[]{ inputWin32, 9, 1, 10 },
                new object[]{ inputWin32, 10, 1, 11 },
                new object[]{ inputWin32, 12, 1, 13 },
                new object[]{ inputWin32, 13, 1, 14 },
                new object[]{ inputWin32, 14, 1, 15 }, // test that encountering \r does not lead to newline
                new object[]{ inputWin32, 15, 2, 1 },  // until \n here on this line.
                new object[]{ inputWin32, 18, 2, 4 },

                new object[]{ inputUnix, 1, 1, 2 },
                new object[]{ inputUnix, 4, 1, 5 },
                new object[]{ inputUnix, 5, 1, 6 },
                new object[]{ inputUnix, 9, 1, 10 },
                new object[]{ inputUnix, 10, 1, 11 },
                new object[]{ inputUnix, 12, 1, 13 },
                new object[]{ inputUnix, 13, 1, 14 },
                new object[]{ inputUnix, 14, 2, 1 },
                new object[]{ inputUnix, 17, 2, 4 },

                new object[]{ inputMac, 1, 1, 2 },
                new object[]{ inputMac, 4, 1, 5 },
                new object[]{ inputMac, 5, 1, 6 },
                new object[]{ inputMac, 9, 1, 10 },
                new object[]{ inputMac, 10, 1, 11 },
                new object[]{ inputMac, 12, 1, 13 },
                new object[]{ inputMac, 13, 1, 14 },
                new object[]{ inputMac, 14, 2, 1 },
                new object[]{ inputMac, 17, 2, 4 },
            };
        }

        [Theory]
        [MemberData(nameof(CreateTestParseOnePartData))]
        public void TestParseOnePart(string source, List<TokenPart> expected)
        {
            LogRecordFormatParser tokenizer = new LogRecordFormatParser(source);
            object part;
            var actual = new List<TokenPart>();
            while ((part = tokenizer.ParseOnePart()) != null)
            {
                Assert.Contains(tokenizer.tokenType, ToList(LogRecordFormatParser.FormatTokenType.LITERAL_STRING_SECTION,
                    LogRecordFormatParser.FormatTokenType.END_REPLACEMENT));
                TokenPart partDescriptor = new TokenPart(part,
                    tokenizer.tokenType == LogRecordFormatParser.FormatTokenType.END_REPLACEMENT,
                    tokenizer.partStart, tokenizer.endPos);
                actual.Add(partDescriptor);
            }
            Assert.Equal(expected, actual);
        }

        public static List<object[]> CreateTestParseOnePartData()
        {
            return new List<object[]>
            {
                new object[]{ "", ToGenericList<TokenPart>() },
                new object[]{ " ", ToGenericList(
                    new TokenPart(" ", false, 0, 1)
                    ) },
                new object[]{ "xyz", ToGenericList(
                    new TokenPart("xyz", false, 0, 3)
                    ) },
                new object[]{ "x{{}}{{yz", ToGenericList(
                    new TokenPart("x{}{yz", false, 0, 9)
                    ) },
                new object[]{ "x{}{{yz", ToGenericList(
                    new TokenPart("x", false, 0, 1),
                    new TokenPart(ToList(), true, 1, 3),
                    new TokenPart("{yz", false, 3, 7)
                    ) },
                new object[]{ "x{0}{{yz", ToGenericList(
                    new TokenPart("x", false, 0, 1),
                    new TokenPart(0, true, 1, 4),
                    new TokenPart("{yz", false, 4, 8)
                    ) },
                new object[]{ "x{a}{{yz", ToGenericList(
                    new TokenPart("x", false, 0, 1),
                    new TokenPart(ToList("a"), true, 1, 4),
                    new TokenPart("{yz", false, 4, 8)
                    ) },
                new object[]{ "x{.0}{{yz", ToGenericList(
                    new TokenPart("x", false, 0, 1),
                    new TokenPart(ToList("0"), true, 1, 5),
                    new TokenPart("{yz", false, 5, 9)
                    ) },

                // test previous 4 again, but with whitespace tolerance.
                new object[]{ "x{ }{{yz", ToGenericList(
                    new TokenPart("x", false, 0, 1),
                    new TokenPart(ToList(), true, 1, 4),
                    new TokenPart("{yz", false, 4, 8)
                    ) },
                new object[]{ "x{ 0 }{{yz", ToGenericList(
                    new TokenPart("x", false, 0, 1),
                    new TokenPart(0, true, 1, 6),
                    new TokenPart("{yz", false, 6, 10)
                    ) },
                new object[]{ "x{a  }{{yz", ToGenericList(
                    new TokenPart("x", false, 0, 1),
                    new TokenPart(ToList("a"), true, 1, 6),
                    new TokenPart("{yz", false, 6, 10)
                    ) },
                new object[]{ "x{ . 0 }{{yz", ToGenericList(
                    new TokenPart("x", false, 0, 1),
                    new TokenPart(ToList("0"), true, 1, 8),
                    new TokenPart("{yz", false, 8, 12)
                    ) },

                // handle longer replacement fields.
                new object[]{ "{bag .prices[0 ]}", ToGenericList(
                    new TokenPart(ToList("bag", "prices", 0), true, 0, 17)
                    ) },
                new object[]{ "{ -12 }{ [10][-2] }{ y.z }", ToGenericList(
                    new TokenPart(-12, true, 0, 7),
                    new TokenPart(ToList(10, -2), true, 7, 19),
                    new TokenPart(ToList("y", "z"), true, 19, 26)
                    ) }
            };
        }
    }
}